### PR TITLE
Automate many more steps of the DTC Ensemble Design Task's workflow

### PR DESCRIPTION
**Pull request by gsketefian**
This pull request enables a more automated way to generate and run the workflows needed for the DTC Ensemble Design task's deliverable of running analogs of the RRFS with IC perturbations and with/without stochastic physics perturbations.

Notes:
1) The original directory containing the scripts for running these workflows (`IC_pert_scripts`, under `ush`) has been renamed `IC_pert_scripts_manual`.
2) The new scripts are in the directory `ICpert_scripts` (note: no underscore after `IC`).
3) The new readme `ICpert_scripts/README` contains detailed instructions on what the new files are and how to run the workflow(s).  Please see that file for details of what this PR does.

### CODE REVIEW

File: scripts/exregional_get_extrn_mdl_files.sh [hunk 1]
```diff
@@ -175,9 +174,95 @@ ${cmd}
...
+to the GEFS file (gefs_fn):
+  gefs_fn = \"${gefs_fn}\"
+  gfs_fn = \"${gfs_fn}\""
+  ln_vrfy -fs --relative "${gefs_fn}" "${gfs_fn}"
+
+fi
+#
+#-----------------------------------------------------------------------
+#
 # Restore the shell options saved at the beginning of this script/function.
```
Comment (JeffBeck-NOAA) [inline scripts/exregional_get_extrn_mdl_files.sh:264]: I'm confused by references to pulling and replacing GFS data above.  The HRRR data is what we're using as the basis for the ICS (prior to adding the GEFS perturbations), so if we're applying the DTC ensemble design task changes, then GET_EXTRN_ICS should have pulled HRRR data, correct?  Also, is there a general high-level switch that enables the ensemble design changes?

File: scripts/exregional_run_fcst.sh [hunk 2]
```diff
@@ -345,8 +345,8 @@ for (( i=0; i<${num_symlinks}; i++ )); do
 
 done
 
-ln_vrfy -sf ${relative_link_flag} ${FIXam}/optics_??.dat ${run_dir}
-ln_vrfy -sf ${relative_link_flag} ${FIXam}/aeroclim.m??.nc ${run_dir}
+#ln_vrfy -sf ${relative_link_flag} ${FIXam}/optics_??.dat ${run_dir}
+#ln_vrfy -sf ${relative_link_flag} ${FIXam}/aeroclim.m??.nc ${run_dir}
 #
 #-----------------------------------------------------------------------
```
Comment (JeffBeck-NOAA) [inline scripts/exregional_run_fcst.sh:348]: Why were these commented out?

File: ush/IC_pert_scripts_manual/README
Comment (JeffBeck-NOAA) [inline ush/IC_pert_scripts_manual/README:1]: Why were things renamed to add "manual"?

File: ush/ICpert_scripts/DTC_ensemble_config_defaults.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+# forecast length of 36 hours, this entails one or more of the following
+# changes:
+#
+# 1) using fewer members
+# 2) using a coarser grid
+# 3) running for fewer cycles
+# 4) using a shorter forecast length
+#
+do_test_run="FALSE"
+#
```
Comment (JeffBeck-NOAA) [inline ush/ICpert_scripts/DTC_ensemble_config_defaults.sh:46]: Are all four options activated above when do_test_run="TRUE"?  If not, how can you select specific options?

File: ush/ICpert_scripts/DTC_ensemble_setup.sh [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+echo
+echo "do_test_run = \"${do_test_run}\""
+echo "RRFS_do_stoch = \"${RRFS_do_stoch}\""
+echo "RRFS_do_get_obs = \"${RRFS_do_get_obs}\""
+echo "GEFS_staging_dir = \"${GEFS_staging_dir}\""
+#
+#-----------------------------------------------------------------------
+#
+# Make sure that required environment variables are defind and, if 
+# applicable, have valid values.
```
Comment (JeffBeck-NOAA) [inline ush/ICpert_scripts/DTC_ensemble_setup.sh:61]: Typo - "defind"

File: ush/ICpert_scripts/DTC_ensemble_setup.sh [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+else
+  RRFS_no_or_null="no"
+fi
+#
+# Absolute paths to the RRFS analog (either without or with stochastic
+# IC perturbations) workflow and to the GEFS individual member workflows.
+#
+RRFS_analog_expt_subdir="RRFS_ICperts_${RRFS_no_or_null}stoch${uscore_test_or_null}"
+RRFS_analog_exptdir="${expt_basedir}/${RRFS_analog_expt_subdir}"
+
```
Comment (JeffBeck-NOAA) [inline ush/ICpert_scripts/DTC_ensemble_setup.sh:144]: What is meant by "analog" here and elsewhere?

File: ush/ICpert_scripts/add_soil_layers.sh [hunk 6]
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
+
+print_info_msg "
+Interpolating soil variables from the 4 layers in NOAH LSM (in GFS output)
```
Comment (JeffBeck-NOAA) [inline ush/ICpert_scripts/add_soil_layers.sh]: While Noah LSM is in GFS, this work is specific to GEFS output, so it would be good to add it.

File: ush/ICpert_scripts/config.sh.GEFS_member [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+PREDEF_GRID_NAME="${predef_grid_name}"
+QUILTING="TRUE"
+HALO_BLEND="20"
+
+CCPP_PHYS_SUITE="FV3_HRRR"
+
+EXTRN_MDL_NAME_ICS="FV3GFS"
+EXTRN_MDL_NAME_LBCS="FV3GFS"
+
```
Comment (JeffBeck-NOAA) [inline ush/ICpert_scripts/config.sh.GEFS_member:29]: Are these set to FV3GFS here simply because we can't pull GEFS data with the SRW App?  I thought maybe we could do that at this point, but maybe not with this version of the App.

File: ush/ICpert_scripts/config.sh.RRFS_analog [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+#
+# If running an ensemble forecast, specify the number of ensemble members
+# and the time lag for each member (ENS_TIME_LAG_HRS).  Also specify 
+# whether to run ensemble vx.
+#
+NUM_ENS_MEMBERS="${num_RRFS_ens_members}"
+ENS_TIME_LAG_HRS=( $( for (( i=0; i<${num_RRFS_ens_members}; i++ )); do echo "0"; done ) )
+#
+# Whether to run deterministic vx.  If running ensemble an forecast, this
+# will cause a deterministic vx to be run on each member.
```
Comment (JeffBeck-NOAA) [inline ush/ICpert_scripts/config.sh.RRFS_analog]: "ensemble an"?

"cause deterministic vx"

File: ush/ICpert_scripts/get_gefs_rrfs.sh [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+#SBATCH --partition=service 
+#SBATCH --job-name=ens_gefs
+#SBATCH --nodes=1
+#SBATCH --ntasks-per-node=1
+#SBATCH --time=5:00:00
+
+#for 30h gefs perturbations to a 00z experiment init, use the 12z init forecasts from 1 day before
+
+set -x
+
```
Comment (JeffBeck-NOAA) [inline ush/ICpert_scripts/get_gefs_rrfs.sh:12]: I see some scripts have "set -x" and others don't.  Is that done on purpose?

File: ush/config_defaults.sh [hunk 10]
```diff
@@ -1399,15 +1399,15 @@ VX_APCP_ACCUMS_HRS=( "01" "03" "06" "24" )
...
 OBS_CCPA_APCPgt01h_FN_TEMPLATE='${OBS_CCPA_APCP01h_FN_TEMPLATE}_a${ACCUM}h.nc'
 
-MRMS_OBS_DIR=""
+MRMS_OBS_DIR='${EXPTDIR}/obs/mrms/proc'
 OBS_MRMS_REFC_FN_TEMPLATE='{valid?fmt=%Y%m%d}/MergedReflectivityQCComposite_00.50_{valid?fmt=%Y%m%d}-{valid?fmt=%H%M%S}.grib2'
 OBS_MRMS_RETOP_FN_TEMPLATE='{valid?fmt=%Y%m%d}/EchoTop_18_00.50_{valid?fmt=%Y%m%d}-{valid?fmt=%H%M%S}.grib2'
 
-NDAS_OBS_DIR=""
+NDAS_OBS_DIR='${EXPTDIR}/obs/ndas/proc'
 OBS_NDAS_SFCorUPA_FN_TEMPLATE='prepbufr.ndas.{valid?fmt=%Y%m%d%H}'
```
Comment (JeffBeck-NOAA) [inline ush/config_defaults.sh:1411]: So these defaults for vx are set here in case someone tells the workflow to pull the obs automatically?  Is this a change that should also go into the authoritative repo?

File: ush/generate_FV3LAM_wflow.sh [hunk 11]
```diff
@@ -402,7 +402,7 @@ file (template_xml_fp):
 #
 # METPlus-specific information
 #
-  'ccpa_obs_dir': ${CCPA_OBS_DIR}
+  'ccpa_obs_dir': $( eval echo ${CCPA_OBS_DIR} )
   'mrms_obs_dir': ${MRMS_OBS_DIR}
   'ndas_obs_dir': ${NDAS_OBS_DIR}
 # Move the variable NET to somewhere more appropriate.
```
Comment (JeffBeck-NOAA) [inline ush/generate_FV3LAM_wflow.sh:408]: Does this change only apply to the CCPA directory and not MRMS or NDAS?

File: ush/generate_FV3LAM_wflow.sh [hunk 12]
```diff
@@ -568,7 +568,8 @@ fi
 #
 #-----------------------------------------------------------------------
 #
-if [ "${RUN_TASK_RUN_FCST}" = "TRUE" ]; then
+if [ "${RUN_TASK_MAKE_OROG}" = "TRUE" ] || \
+   [ "${RUN_TASK_RUN_FCST}" = "TRUE" ]; then
 #
```
Comment (JeffBeck-NOAA) [inline ush/generate_FV3LAM_wflow.sh:573]: Why is this change needed?

File: ush/setup.sh [hunk 13]
```diff
@@ -1285,9 +1285,11 @@ The CCPP physics suite specified in CCPP_PHYS_SUITE is not supported:
...
     "CCPP_PHYS_SUITE" "valid_vals_CCPP_PHYS_SUITE" "${err_msg}"
 #
-# If running the RUN_FCST_TN task...
+# If running the MAKE_ICS_TN, MAKE_LBCS_TN, or RUN_FCST_TN task...
 #
-  if [ "${RUN_TASK_RUN_FCST}" = "TRUE" ]; then
+  if [ "${RUN_TASK_MAKE_ICS}" = "TRUE" ] || \
+     [ "${RUN_TASK_MAKE_LBCS}" = "TRUE" ] || \
+     [ "${RUN_TASK_RUN_FCST}" = "TRUE" ]; then
 #
```
Comment (JeffBeck-NOAA) [inline ush/setup.sh:1293]: Why include MAKE_ICS and MAKE_LBCS in this "if" statement?

File: scripts/exregional_run_fcst.sh (hunk 2 above)
Comment (gsketefian) [reply scripts/exregional_run_fcst.sh:348]: Because they don't work and the loop below (inside `IF ${USE_MERRA_CLIMO}...`) already does what needs to be done.  @willmayfield didn't realize at the time that setting `USE_MERRA_CLIMO="TRUE"` already does what he's trying to do.  Actually, I think it's best to just remove these lines to avoid confusion.

File: ush/IC_pert_scripts_manual/README
Comment (gsketefian) [reply ush/IC_pert_scripts_manual/README:1]: I didn't want to just remove all of @willmayfield 's scripts since he may want to have them around to compare them to the new more automated scripts in `ICpert_scripts`.  So I renamed them, emphasizing that they are the more manual way of doing things.  In addition, there was a commit by @briannen last week (of scripts to get the GEFS files) that should be included in `IC_pert_scripts_manual`, another reason to keep the old script versions around.

File: ush/ICpert_scripts/DTC_ensemble_config_defaults.sh (hunk 3 above)
Comment (gsketefian) [reply ush/ICpert_scripts/DTC_ensemble_config_defaults.sh:46]: Yes, but you can also pick and choose which ones you want to override.  Note that this file contains default values (much like `config_defaults.sh` in the SRW App), and the user can include a user configure file named `DTC_ensemble_config.sh` that can override the default values.  The `README` describes this in more detail.

File: ush/ICpert_scripts/DTC_ensemble_setup.sh (hunk 4 above)
Comment (gsketefian) [reply ush/ICpert_scripts/DTC_ensemble_setup.sh:61]: Fixed.  Will push soon.

File: ush/ICpert_scripts/DTC_ensemble_setup.sh (hunk 5 above)
Comment (gsketefian) [reply ush/ICpert_scripts/DTC_ensemble_setup.sh:144]: Our RRFS runs are analogs of the real RRFS runs that GSL does.  I saw @willmayfield use this terminology, so I used it too.  It's to distinguish between the GEFS individual member workflows and the RRFS workflows (one nostoch, another stoch) that we're running.

File: ush/ICpert_scripts/add_soil_layers.sh (hunk 6 above)
Comment (gsketefian) [reply ush/ICpert_scripts/add_soil_layers.sh]: Ok, I edited this for clarity.  Take a look at the new version once I push and let me know what you think.

File: ush/ICpert_scripts/config.sh.GEFS_member (hunk 7 above)
Comment (gsketefian) [reply ush/ICpert_scripts/config.sh.GEFS_member:29]: I wondered about that but didn't want to explore it for lack of time and we know that this works with @willmayfield 's scripts.  If that does work in this version, we could just set the ICs one to GEFS, but I haven't explored this.

File: ush/ICpert_scripts/config.sh.RRFS_analog (hunk 8 above)
Comment (gsketefian) [reply ush/ICpert_scripts/config.sh.RRFS_analog]: Ok, edited.  Let me know what you think after I push the latest.

File: ush/ICpert_scripts/get_gefs_rrfs.sh (hunk 9 above)
Comment (gsketefian) [reply ush/ICpert_scripts/get_gefs_rrfs.sh:12]: No.  This is not one of the scripts I worked on.  I think @briannen committed this last week, and she chose to include this to get more debugging info.

File: ush/config_defaults.sh (hunk 10 above)
Comment (gsketefian) [reply ush/config_defaults.sh:1411]: Q1:  Yes.
Q2:  I think I included something like this in one my PRs into the authoritative repo's developed branch, but that stuff has now dragged on so long I've lost track of where things are.

File: scripts/exregional_run_fcst.sh (hunk 2 above)
Comment (JeffBeck-NOAA) [reply scripts/exregional_run_fcst.sh:348]: Thanks!  Yes, I'd agree that it's best just to remove them.

File: ush/IC_pert_scripts_manual/README
Comment (JeffBeck-NOAA) [reply ush/IC_pert_scripts_manual/README:1]: This makes sense, thanks!

File: ush/generate_FV3LAM_wflow.sh (hunk 11 above)
Comment (gsketefian) [reply ush/generate_FV3LAM_wflow.sh:408]: Yeah, I think I should apply to the other two as well.  Will fix.  But it only matters if the `GET_OBS_...` tasks are activated, which they are not in all our experiments.

File: ush/ICpert_scripts/DTC_ensemble_config_defaults.sh (hunk 3 above)
Comment (JeffBeck-NOAA) [reply ush/ICpert_scripts/DTC_ensemble_config_defaults.sh:46]: Oh, whoops, I missed that this was config_defaults.sh and not config.sh.  OK, thank you!

File: ush/ICpert_scripts/config.sh.GEFS_member (hunk 7 above)
Comment (JeffBeck-NOAA) [reply ush/ICpert_scripts/config.sh.GEFS_member:29]: OK, no problem.  It works as is!

File: ush/ICpert_scripts/get_gefs_rrfs.sh (hunk 9 above)
Comment (JeffBeck-NOAA) [reply ush/ICpert_scripts/get_gefs_rrfs.sh:12]: Got it.  Thanks!

File: ush/config_defaults.sh (hunk 10 above)
Comment (JeffBeck-NOAA) [reply ush/config_defaults.sh:1411]: OK, no problem.  Thanks!

File: ush/generate_FV3LAM_wflow.sh (hunk 11 above)
Comment (JeffBeck-NOAA) [reply ush/generate_FV3LAM_wflow.sh:408]: Yep, good point.  We'll have everything staged.

File: ush/generate_FV3LAM_wflow.sh (hunk 12 above)
Comment (gsketefian) [reply ush/generate_FV3LAM_wflow.sh:573]: Because when running `make_orog` but not `make_fcst`, it turns out the `FIXam` directory is still needed.  I think we had never tested the App in that configuration before.  And yes, something like this is probably also needed in the authoritative repo's `develop` branch, but again I've lost track and don't know if it's already in there or in one of the PRs that are slated to be merged.

File: ush/setup.sh (hunk 13 above)
Comment (gsketefian) [reply ush/setup.sh:1293]: Like above, this is a bug fix to ensure that the variables that reference the CCPP suite definition file are properly defined if running the `MAKE_ICS` or `MAKE_LBCS` task but not the `RUN_FCST` task.  Again, apparently a configuration that wasn't tested, or maybe this has already been fixed in `develop`.

File: ush/generate_FV3LAM_wflow.sh (hunk 12 above)
Comment (JeffBeck-NOAA) [reply ush/generate_FV3LAM_wflow.sh:573]: Ahh, interesting.  Yep, that was going to be my next question (how is this working in the authoritative repo?).  I would assume the develop branch would be broken without the FIXam directory for make_orog, since that's the case here, so I imagine it's been modified.

File: ush/setup.sh (hunk 13 above)
Comment (JeffBeck-NOAA) [reply ush/setup.sh:1293]: Got it!  Thanks!  Yes, I doubt this configuration was ever tested in the App before (run chgres_cube, but not the model).

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 1 above)
Comment (gsketefian) [reply scripts/exregional_get_extrn_mdl_files.sh:264]: Remember that these need to generate both an RRFS workflow and a GEFS workflow.  This new code section will only get activated when running a GEFS workflow (for a GEFS member).  For an RRFS workflow, the variable `gefs_mem` set on line ~191 will be an empty string because the experiment name will not have the form `GEFS_mem##` (the experiment name will be something like `RRFS_ICperts_[no]stoch`).  Thus, the condition in the if-statement on line ~205 will not true, and the code in it will not be executed (so the `HRRR` files obtained above will not be replaced).  But for a GEFS workflow, `gefs_mem` will be `01` or `02` or ..., so the if-statement will be true, and the GFS external model file will be replaced with the staged GEFS file.

File: ush/IC_pert_scripts_manual/README
Comment (willmayfield) [reply ush/IC_pert_scripts_manual/README:1]: I'm going to make an update to organize the get_gefs... scripts after this PR goes in. At that time I can also remove the _manual directory.

File: scripts/exregional_get_extrn_mdl_files.sh (hunk 1 above)
Comment (JeffBeck-NOAA) [reply scripts/exregional_get_extrn_mdl_files.sh:264]: Thanks for the explanation!

**Review by JeffBeck-NOAA — APPROVED** [review]

File: ush/ICpert_scripts/DTC_ensemble_setup.sh (hunk 5 above)
Comment (willmayfield) [reply ush/ICpert_scripts/DTC_ensemble_setup.sh:144]: I think I picked up that usage of analog from Jeff.

**Review by willmayfield — APPROVED** [review]
